packages/eservice-template-process/src/services/readModelServiceSQL.ts

```diff
@@ -0,0 +1,373 @@
+      return (await tenantReadModelServiceSQL.getTenantById(id))?.data;
+    },
+    async getAttributesByIds(
+      attributesIds: AttributeId[]
+    ): Promise<Attribute[]> {
+      return (
+        await attributeReadModelServiceSQL.getAttributesByFilter(
+          inArray(attributeInReadmodelAttribute.id, attributesIds)
+        )
+      ).map((a) => a.data);
+    },
+    async getEServiceTemplates(
+      filters: GetEServiceTemplatesFilters,
+      offset: number,
+      limit: number,
+      authData: UIAuthData | M2MAuthData
+    ): Promise<ListResult<EServiceTemplate>> {
+      const { eserviceTemplatesIds, creatorsIds, states, name } = filters;
+
+      const subquery = readModelDB
+        .select({
+          eserviceTemplateId: eserviceTemplateInReadmodelEserviceTemplate.id,
+          totalCount: sql`COUNT(*) OVER()`.mapWith(Number).as("totalCount"),
+        })
+        .from(eserviceTemplateInReadmodelEserviceTemplate)
+        .leftJoin(
+          eserviceTemplateVersionInReadmodelEserviceTemplate,
+          eq(
+            eserviceTemplateInReadmodelEserviceTemplate.id,
+            eserviceTemplateVersionInReadmodelEserviceTemplate.eserviceTemplateId
+          )
+        )
+        .where(
+          and(
+            // NAME FILTER
+            name
+              ? ilike(
+                  eserviceTemplateInReadmodelEserviceTemplate.name,
+                  `%${ReadModelRepository.escapeRegExp(name)}%`
+                )
+              : undefined,
+            // IDS FILTER
+            eserviceTemplatesIds.length > 0
+              ? inArray(
+                  eserviceTemplateInReadmodelEserviceTemplate.id,
+                  eserviceTemplatesIds
+                )
+              : undefined,
+            // CREATORS IDS FILTER
+            creatorsIds.length > 0
+              ? inArray(
+                  eserviceTemplateInReadmodelEserviceTemplate.creatorId,
+                  creatorsIds
+                )
+              : undefined,
+            // STATES FILTER
+            states.length > 0
+              ? inArray(
+                  eserviceTemplateVersionInReadmodelEserviceTemplate.state,
+                  states
+                )
+              : undefined,
+            // VISIBILITY FILTER
+            or(
+              hasAtLeastOneUserRole(authData, [
+                userRole.ADMIN_ROLE,
+                userRole.API_ROLE,
+                userRole.SUPPORT_ROLE,
+              ])
+                ? eq(
+                    eserviceTemplateInReadmodelEserviceTemplate.creatorId,
+                    authData.organizationId
+                  )
+                : undefined,
+              and(
+                ne(
+                  eserviceTemplateVersionInReadmodelEserviceTemplate.state,
+                  eserviceTemplateVersionState.draft
+                ),
+                isNotNull(eserviceTemplateVersionInReadmodelEserviceTemplate.id)
+              )
+            )
```

Comment:
This looks different from the filter in the old readmodel - the filter on the existing version + the non-draft check was applied in both cases, no matter the user roles, so it should probably go outside this `or` and become a check in `and` with the rest. Here you can leave the check on the creatorId being the requester, only in case the user role is admin / API / support

This is the filter I'm talking about 👇 
```ts
                { "data.versions.1": { $exists: true } },
                {
                  "data.versions": { $size: 1 },
                  "data.versions.0.state": {
                    $ne: eserviceTemplateVersionState.draft,
                  },
                }
```

Reply:
I don't get it. Both [these](https://github.com/pagopa/interop-be-monorepo/blob/develop/packages/eservice-template-process/src/services/readModelService.ts#L183-L204) visibility conditions are in `or`s. The only different part is `{ "data.creatorId": authData.organizationId }` which is satisfied by
```
              hasAtLeastOneUserRole(authData, [
                userRole.ADMIN_ROLE,
                userRole.API_ROLE,
                userRole.SUPPORT_ROLE,
              ])
                ? eq(
                    eserviceTemplateInReadmodelEserviceTemplate.creatorId,
                    authData.organizationId
                  )
                : undefined,
```
The rest
```
              and(
                ne(
                  eserviceTemplateVersionInReadmodelEserviceTemplate.state,
                  eserviceTemplateVersionState.draft
                ),
                isNotNull(eserviceTemplateVersionInReadmodelEserviceTemplate.id)
              )
```
 is the translation of
```
                 { "data.versions.1": { $exists: true } },
                {
                  "data.versions": { $size: 1 },
                  "data.versions.0.state": {
                    $ne: eserviceTemplateVersionState.draft,
                  },
                }
```
So, if `hasAtLeastOneUserRole` is false it ignores that condition and the whole SQL `or(...)` becomes [this](https://github.com/pagopa/interop-be-monorepo/blob/develop/packages/eservice-template-process/src/services/readModelService.ts#L195-L203). Am I missing something?

The only difference should be that the old query would've gotten e-service templates with versions `[draft, draft]` and the new one does not. Though, I don't think this case is allowed by the platform.

Reply:
I think ` { "data.versions.1": { $exists: true } },` allowed the retrieval for any template that has at least two versions, no matter the state of versions, so the last version could also be still draft e.g., `[published, draft]`, `[published, published, draft]` -- would this new code return these?  If yes, then let's leave it like this :)

Still, I would prefer to keep the porting in the same way it was coded before, to replicate the exact same behavior. As we did in a similar condition in the catalog, where we kept the or with the duplicated check in both two cases of the or https://github.com/pagopa/interop-be-monorepo/pull/1683/files#diff-f1a4d3f3bd5a0eceef0331d4a8717681633f8e9be4e0f5cb6a534a82120ceab5R224

Reply:
@shuyec 🆙

Reply:
I also agree that this isn't exactly equivalent to the old one, I would keep it like this
```suggestion
            or(
              and(
                hasAtLeastOneUserRole(authData, [
                  userRole.ADMIN_ROLE,
                  userRole.API_ROLE,
                  userRole.SUPPORT_ROLE,
                ]),
                eq(
                  eserviceTemplateInReadmodelEserviceTemplate.creatorId,
                  authData.organizationId
                )
              ),
              and(
                ne(
                  eserviceTemplateVersionInReadmodelEserviceTemplate.state,
                  eserviceTemplateVersionState.draft
                ),
                isNotNull(eserviceTemplateVersionInReadmodelEserviceTemplate.id)
              )
            )
```

Reply:
We [refactored](https://github.com/pagopa/interop-be-monorepo/pull/1742/commits/be0a9119b4573d6dfccac43825e7d463991f1c65) that with a 1:1 translation from the Mongo query because of clarity.
We also checked that the old version would've worked anyway because Drizzle completely ignores conditions that are `undefined` (they're not present in the SQL translation), so `eserviceTemplateInReadmodelEserviceTemplate.creatorId === authData.organizationId` would've been ignored with `hasAtLeastOneUserRole === false`.